bio_hansel/subtyper.py

```diff
@@ -228,6 +228,30 @@ def parallel_query_reads(reads: List[Tuple[List[str], str]],
     outputs = [x.get() for x in res]
     return outputs
 
+def calc_kmer_fraction(df,min_kmer_frac=0.05):
+    """Calculate the percentage each k-mer frequence represents for a given position
+
+    Args:
+        df: BioHansel k-mer frequence pandas df
+        min_kmer_frac: float 0 - 1 on the minimum fraction a kmer needs to be to be considered valid
+
+    Returns:
+        - pd.DataFrame with k-mers with kmer_fraction column
+    """
+    position_frequencies = df[['refposition','freq']].groupby(['refposition']).sum().reset_index()
+    percentages = []
+    total_refposition_kmer_frequencies = []
+    for index,row in df.iterrows():
```

Comment:
I'd recommend using `.itertuples()` for performance reasons:

```python
for row in df.itertuples():
    refposition = row.refposition # cannot do string based access (row['refposition']) with tuples
```

You could also look into using [apply](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.apply.html) instead of using a for-loop:

```python
def get_kmer_fraction(row):
    total_freq = position_frequencies.get(row.refposition, 0)
    return row.freq / total_freq if total_freq > 0 else 0.0

df['kmer_fraction'] = df.apply(get_kmer_fraction, axis=1)
df['total_refposition_kmer_frequency'] = df.apply(lambda row: position_frequencies.get(row.refposition, 0), axis=1) 
```